package/MDAnalysis/core/topologyattrs.py

```diff
@@ -2032,6 +2045,239 @@ def total_charge(group, compound='group'):
+                dipole_moment[compound_mask] = np.sum(
+                    (coords[atom_mask]-ref[compound_mask][:,None,:])
+                    *chgs[atom_mask][:,:,None], axis=1)
+
+        return dipole_moment
+
+    transplants[GroupBase].append(
+        ('dipole_moment', dipole_moment))
+
+    @warn_if_not_unique
+    @_pbc_to_wrap
+    @check_wrap_and_unwrap
+    def quadrupole_moment(group, wrap=False, unwrap=False, 
+        compound='group', center="mass"):
+        r"""Quadrupole vector of the group according to [Gray1984]_
+         
+        .. math::
+            Q = \sqrt{\frac{2}{3}\boldsymbol{\hat{Q}}:\boldsymbol{\hat{Q}}}
+
+        where :math:`\boldsymbol{\hat{Q}}` is the traceless quadrupole 
+        tensor.
+
+        Computes the quadrupole moment of :class:`Atoms<Atom>` in the group.
+        Dipoleper :class:`Residue`, :class:`Segment`, molecule, or
+        fragment can be obtained by setting the `compound` parameter
+        accordingly.
+
+        Parameters
+        ----------
+        group : obj
+            AtomGroup
+        wrap : bool, optional
+            If ``True`` and `compound` is ``'group'``, move all atoms to the
+            primary unit cell before calculation.
+            If ``True`` and `compound` is not ``group``, the
+            centers of mass of each compound will be calculated without moving
+            any atoms to keep the compounds intact.
+        unwrap : bool, optional
+            If ``True``, compounds will be unwrapped before computing their
+            centers.
+        compound : {'group', 'segments', 'residues', 'molecules', \
+                    'fragments'}, optional
+            If ``'group'``, a single quadrupole value returns. Otherwise, an
+            array of each :class:`Segment`, :class:`Residue`, molecule, or
+            fragment will be returned as an array of position vectors, i.e.
+            a 1d array.
+            Note that, in any case, *only* the positions of
+            :class:`Atoms<Atom>` *belonging to the group* will be taken into
+            account.
+        center : str, optional
+            Choose whether the quadrupole moment is calculated at the center of 
+            "mass" or the center of "charge"
+
+        Returns
+        -------
+        numpy.ndarray
+            Quadrupole moment of (compounds of) the group in e*angstroms**2.
+            If `compound` was set to ``'group'``, the output will be a single
+            value. Otherwise, the output will be a 1d array of shape ``(n,)``
+            where ``n`` is the number of compounds.
+
+        .. versionadded:: 2.4.0
+        """
+        def __quadrupole(recenteredpos, charges):
+            tensor = np.zeros((3,3))
+            for i,atom in enumerate(recenteredpos):
+                tensor += np.matmul(
+                    atom[:,np.newaxis],
+                    atom[np.newaxis,:])*charges[i]
+
+            quad_trace = np.sum(np.diag(tensor))
```

Comment:
Nitpick but  think you could possibly use [`np.trace` ](https://numpy.org/doc/stable/reference/generated/numpy.trace.html)